### PR TITLE
[Core] enhance compare process

### PR DESCRIPTION
**Pull request by philbucher**
this PR merges functionalities from two comparison processes

I need this for sth I am working on at the moment and also for further tests in mpi (where we cannot do the tests based on the node number bcs partitioning reorders them)

@loumalouomega I did a bit of reorganizaition but nothing major, the tests in the Meshingapp should still work
Now everything is quite consistent

@msandre I merged the processes written by you into the other one, I adapted all your tests, all the tests in AdjointFluid are working, also the mpi test. 
I hope this is ok for you

### CODE REVIEW

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 1]
```diff
@@ -139,52 +152,54 @@ def __CompareResultsBlock(self, lines1, lines2, current_index):
...
                 break
 
         error = 0.0
         for i in range(numline, nvertices + numline):
-            tmp1 = ConvertStringToListFloat(lines1[i], "", "\n")
-            tmp2 = ConvertStringToListFloat(lines2[i], "", "\n")
+            tmp1 = ConvertStringToListFloat(lines_ref[i], "", "\n")
+            tmp2 = ConvertStringToListFloat(lines_out[i], "", "\n")
             if (dimension == 2):
                 error += ((tmp1[0] - tmp2[0])**2.0 + (tmp1[1] - tmp2[1])**2.0)**(0.5)
```
Comment (RiccardoRossi) [inline kratos/python_scripts/compare_two_files_check_process.py]: guys i have a serious problem to understand why you use a double exponent 2.0 

I am very convinced that you should use 2 (not 2.0 as exponent). Also DO NOT use exponent 0.5 for sqrt, use the sqrt instead which goes through a specialized pathway!!

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 2]
```diff
@@ -59,56 +59,69 @@ def ExecuteAfterOutputStep(self):
...
-        elif (self.comparison_type == "post_res_file"):
-            self.__ComparePostResFile()
-        else:
-            raise NameError('Requested comparision type "' + self.comparison_type + '" not implemented yet')
+        if kratos_utils.IsRankZero():
+            if (self.comparison_type == "deterministic"):
+                value = filecmp.cmp(self.reference_file_name, self.output_file_name)
+                self.assertTrue(value)
+            elif (self.comparison_type == "mesh_file"):
+                self.__CompareMeshVertivesFile()
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: typo

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 3]
```diff
@@ -59,56 +59,69 @@ def ExecuteAfterOutputStep(self):
...
-            for i in range(num_lines_1):
-                if lines_ref[i].startswith("Result "):
-                    results_start_index = i
-                    results_found = True
-                    break
+        return lines_ref, lines_out
 
-                lines_ref_splitted = lines_ref[i].split()
-                lines_out_splitted = lines_out[i].split()
+    def __ComparePostResFile(self):
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: A docstring explaining what a PostResFile is (i.e. a reference or a description of it's format) should be included.

File: kratos/python_scripts/compare_two_files_check_process.py (hunk 1 above)
Comment (philbucher) [reply kratos/python_scripts/compare_two_files_check_process.py]: makes sense to me, this was written by @loumalouomega 

@loumalouomega I will change it if you are ok with it

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 4]
```diff
@@ -59,56 +59,69 @@ def ExecuteAfterOutputStep(self):
...
-                    self.assertTrue(ref_value == out_value,
-                                    msg=ref_value + " != " + out_value)
+            lines_ref_splitted = lines_ref[i].split()
+            lines_out_splitted = lines_out[i].split()
 
-            # comparing the results
-            if results_found:
-                while results_start_index < num_lines_1:
-                    results_start_index = self.__CompareResultsBlock(lines_ref, lines_out, results_start_index)
+            if len(lines_ref_splitted) != len(lines_out_splitted):
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: The comparison can be embedded in the assert function.

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 5]
```diff
@@ -59,56 +59,69 @@ def ExecuteAfterOutputStep(self):
...
+        lines_ref, lines_out = self.__GetFileLines()
 
-                if len(lines_ref_splitted) != len(lines_out_splitted):
-                    self.assertTrue(False, msg="Lines have different length!")
+        results_start_index = -1
+        results_found = False
+
+        num_lines = len(lines_ref)
+
+        # comparing the header
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: This function is long and complex. Please refactor "comparing the header" to a new function.

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 6]
```diff
@@ -139,52 +152,54 @@ def __CompareResultsBlock(self, lines1, lines2, current_index):
...
+            for v1, v2 in zip(line_ref.split(), line_out.split()):
+                self.assertAlmostEqual(float(v1),
+                                       float(v2),
+                                       self.decimal_places)
 
-def _ReadVertices(input_file1, input_file2, dimension):
-    with open(input_file1,'r') as f1, open(input_file2,'r') as f2:
-        lines1 = f1.readlines()
-        lines2 = f2.readlines()
+    def __CompareMeshVertivesFile(self):
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: typo

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 7]
```diff
@@ -139,52 +152,54 @@ def __CompareResultsBlock(self, lines1, lines2, current_index):
 
         return current_index+2 # directly incrementing to get the new result label
 
+    def __CompareDatFile(self):
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py:181]: This function follows a similar policy as the previous with some differences in low level detail. Step 1 - compare header lines for equality, Step 2 - compare floating point numbers with almost equal. The code would be simpler to read and probably easier to maintain if you extract the low level operations to separate functions. When reading the __Compare*File() functions one should quickly read the higher level stages without getting slowed down by the low level details.

File: kratos/python_scripts/compare_two_files_check_process.py (hunk 6 above)
Comment (msandre) [reply kratos/python_scripts/compare_two_files_check_process.py]: Also a docstring referencing the format of such file would be helpful.

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 8]
```diff
@@ -139,52 +152,54 @@ def __CompareResultsBlock(self, lines1, lines2, current_index):
...
-    return (error/nvertices)
+        error /= nvertices
+        self.assertTrue(error < GetTolerance(self.decimal_places))
 
-def ReadMetric(input_file1, input_file2, dimension):
-    with open(input_file1,'r') as f1, open(input_file2,'r') as f2:
-        lines1 = f1.readlines()
-        lines2 = f2.readlines()
+    def __CompareSolMetricFile(self):
+        lines_ref, lines_out = self.__GetFileLines()
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: docstring please. I don't know what a SolMetricFile is.

File: kratos/python_scripts/compare_two_files_check_process.py [hunk 9]
```diff
@@ -199,19 +214,29 @@ def ReadMetric(input_file1, input_file2, dimension):
...
-            tmp2 = ConvertStringToListFloat(lines2[i], space, end_line)
+            if (lines_ref[i][0] == " "):
+                lines_ref[i] = lines_ref[i][1:]
+            if (lines_out[i][0] == " "):
+                lines_ref[i][0] = lines_out[i][1:]
+            tmp1 = ConvertStringToListFloat(lines_ref[i], space, end_line)
+            tmp2 = ConvertStringToListFloat(lines_out[i], space, end_line)
 
             if (dimension == 2):
                 error += ((tmp1[0] - tmp2[0])**2.0 + (tmp1[1] - tmp2[1])**2.0 + (tmp1[2] - tmp2[2])**2.0)**(0.5)
```
Comment (msandre) [inline kratos/python_scripts/compare_two_files_check_process.py]: Add this to @RiccardoRossi 's comment above.

**Review by msandre — CHANGES_REQUESTED** [review]
Thanks for checking that the tests work. I just have some comments mostly about the code organization to make it easier to understand and maintain.

File: kratos/python_scripts/compare_two_files_check_process.py (hunk 8 above)
Comment (loumalouomega) [reply kratos/python_scripts/compare_two_files_check_process.py]: Sorry, it is a specific output/input file for the MMG library

**Review by loumalouomega — DISMISSED** [review]
Apply changes suggested by @msandre  and @RiccardoRossi please

**Comment by philbucher** [comment]
@msandre the tests in AdointFluid produce a lot of output in the terminal, I would recommend reducing this for a better overview

**Comment by philbucher** [comment]
I should have addressed all comments

Documentation is added, please let me know if it is ok / sufficient

**Review by msandre — APPROVED** [review]
Approving. @philbucher I think you mean the relaxation output in the adjoint tests. This is a small effort to fix and I will try to do it the next time we clean up the adjoint tests.

**Comment by philbucher** [comment]
@msandre yes exactly

You could use the Logger and increase the Verbosity lvl in the tests
This is done in StructuralMechanics, I can explain it to you if you are interested

**Pull request merged**